codesigning/gcp-cloudfunc/main.py

```diff
@@ -0,0 +1,175 @@
+import contextlib
+import datetime
+import json
+import logging
+import os
+import time
+from urllib.parse import unquote
+
+import functions_framework
+import google.cloud.logging  # pip install google-cloud-logging
+import requests
+from flask import jsonify
+from google.cloud import storage  # pip install google-cloud-storage
+
+GOOGLE_PREFIX = 'https://storage.googleapis.com'
+CODESIGN_DATA_BUCKET = 'natcap-codesigning'
+LOG_CLIENT = google.cloud.logging.Client()
+LOG_CLIENT.setup_logging()
+
+
+@contextlib.contextmanager
+def get_lock():
+    """Acquire a GCS-based mutex.
+
+    This requires that the bucket we are using has versioning.
+    """
+    storage_client = storage.Client()
+    bucket = storage_client.bucket(CODESIGN_DATA_BUCKET)
+
+    lock_obtained = False
+    n_tries = 100
+    for i in range(n_tries):
+        lockfile = bucket.blob('mutex.lock')
+        if not lockfile.generation:
+            lockfile.upload_from_string(
+                f"Lock acquired {datetime.datetime.now().isoformat()}")
+            lock_obtained = True
+            break
+        else:
+            time.sleep(0.1)
+
+    if not lock_obtained:
+        raise RuntimeError(f'Could not obtain lock after {n_tries} tries')
+
+    try:
+        yield
+    finally:
+        lockfile.delete()
+
+
+@functions_framework.http
+def main(request):
+    """Handle requests to this GCP Cloud Function.
+
+    All requests must be POST requests and have a JSON body with the following
+    attributes:
+
+        * token: a secret token that matches the ACCESS_TOKEN environment
+            variable that is defined in the cloud function configuration.
+        * action: either 'enqueue' or 'dequeue'
+
+    If the action is 'enqueue', the request must also have a 'url' attribute.
+    """
+    data = request.get_json()
+    if data['token'] != os.environ['ACCESS_TOKEN']:
+        logging.info('Rejecting request due to invalid token')
+        return jsonify('Invalid token'), 403
+
+    if request.method != 'POST':
+        logging.info('Rejecting request due to invalid HTTP method')
+        return jsonify('Invalid request method'), 405
+
+    storage_client = storage.Client()
+    bucket = storage_client.bucket(CODESIGN_DATA_BUCKET)
+
+    logging.debug('Data POSTed: %s', data)
```

Comment:
I'd suggest the modern f-string style
```suggestion
    logging.debug(f'Data POSTed: {data}')
```